server/src/rate_limit/mod.rs

```diff
@@ -2,17 +2,182 @@ pub mod rate_limiter;
 
 use super::{IPAddr, Settings};
 use crate::api::APIError;
+use crate::settings::RateLimitConfig;
+use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
 use failure::Error;
+use futures::future::{ok, Ready};
 use log::warn;
-use rate_limiter::RateLimiter;
+use rate_limiter::{RateLimitType, RateLimiter};
 use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
-use std::sync::Mutex;
+use std::task::{Context, Poll};
 use std::time::SystemTime;
 use strum::IntoEnumIterator;
+use tokio::sync::Mutex;
 
 #[derive(Debug, Clone)]
-pub struct RateLimitInfo {
-  pub rate_limiter: Arc<Mutex<RateLimiter>>,
-  pub ip: IPAddr,
+pub struct RateLimit(pub Arc<Mutex<RateLimiter>>);
+
+#[derive(Debug, Clone)]
+pub struct RateLimited(Arc<Mutex<RateLimiter>>, RateLimitType);
+
+pub struct RateLimitedMiddleware<S>(RateLimited, S);
+
+impl RateLimit {
+  pub fn message(&self) -> RateLimited {
+    self.kind(RateLimitType::Message)
+  }
+
+  pub fn post(&self) -> RateLimited {
+    self.kind(RateLimitType::Post)
+  }
+
+  pub fn register(&self) -> RateLimited {
+    self.kind(RateLimitType::Register)
+  }
+
+  fn kind(&self, type_: RateLimitType) -> RateLimited {
+    RateLimited(self.0.clone(), type_)
+  }
+}
+
+impl RateLimited {
+  pub async fn wrap<T, E>(
+    self,
+    ip_addr: String,
+    fut: impl Future<Output = Result<T, E>>,
+  ) -> Result<T, E>
+  where
+    E: From<failure::Error>,
+  {
+    let rate_limit: RateLimitConfig = actix_web::web::block(move || {
+      // needs to be in a web::block because the RwLock in settings is from stdlib
+      Ok(Settings::get().rate_limit) as Result<_, failure::Error>
```

Comment:
At some point I wanna move the settings from a singleton lazy_static to a state mutex, just like this.